cores/esp32/HardwareSerial.cpp

```diff
@@ -120,24 +128,44 @@ void HardwareSerial::begin(unsigned long baud, uint32_t config, int8_t rxPin, in
         // thus do not disable debug output
         end(false);
     }
-    if(_uart_nr == 0 && rxPin < 0 && txPin < 0) {
-        rxPin = SOC_RX0;
-        txPin = SOC_TX0;
+
+    if (rxPin > -1) {
+        _rxPin = rxPin;
+    }
+    if (txPin > -1) {
+        _txPin = txPin;
+    }
+
+    if(_uart_nr == 0) {
+        if (_rxPin < 0) {
+            _rxPin = SOC_RX0;
+        }
+        if (_txPin < 0) {
+            _txPin = SOC_TX0;
+        }
     }
 #if SOC_UART_NUM > 1
-    if(_uart_nr == 1 && rxPin < 0 && txPin < 0) {
-        rxPin = RX1;
-        txPin = TX1;
+    if(_uart_nr == 1) {
```

Comment:
those do the same exact thing as before :) reassign the pins if they are not `-1`

Reply:
Nope, they assign to the _rxPin and _txPin, which are now protected members of the class, so they persist.
They are now only changed in begin() if:
- rxPin parameter is > -1 -> we take its value as the _rxPin value
- txPin parameter is > -1 -> we take its value as the _txPin value

Then later, if any of the _rxPin/_txPin is still -1 we take the defaults.

This allows us to only override the already assigned pins where actually needed I believe..